### ISSUE TITLE
Inferred type in generic function merges/unions all supplied types

### ISSUE THREAD

**Issue by andrewrosss** [microsoft/pyright#3759]
**Describe the bug**

A generic function with multiple arguments using the same generic parameter results in the inferred type of the function arguments being the union of all types passed to the corresponding arguments.

**To Reproduce**

```python
T = TypeVar("T")
U = TypeVar("U")


def f(t: T, u: U, cb: Callable[[T], U]) -> Any:
    ...


def cb(x: int) -> bool:
    return x > 0


f(t=1, u="a", cb=cb)  # No error here - expected an error
```

**Expected behavior**

I seems like the above code should results in a type error. One of the following two possibilities makes sense depending on the order that the types are inferred, either:

- A type error related to `cb`: `u` is inferred to be of type `str`, so `cb` should be of type `Callable[[int], str]`, but the supplied function is of type `Callable[int, bool]`.
- A type error related to `u`: `cb` is inferred to be of type `Callable[int, bool]`, so `u` should be of type `bool`, but the supplied argument is of type `str`.

For example, running `mypy` (v0.971) on this code results in the first error:

```bash
$ mypy example.py 
example.py:18: error: Argument "cb" to "f" has incompatible type "Callable[[int], bool]"; expected "Callable[[int], str]"
Found 1 error in 1 file (checked 1 source file)
```

**Screenshots or Code**

<img width="664" alt="Screen Shot 2022-07-28 at 5 21 17 PM" src="https://user-images.githubusercontent.com/23389447/181639841-71f6e9d2-f2e6-4cb9-81e1-58c5fcb45377.png">

**VS Code extension or command-line**

Using Pyright in VSCode via the Pylance (v2022.7.40) extension

About VSCode:
```
Version: 1.69.2 (Universal)
Commit: 3b889b090b5ad5793f524b5d1d39fda662b96a2a
Date: 2022-07-18T16:12:57.074Z (1 wk ago)
Electron: 18.3.5
Chromium: 100.0.4896.160
Node.js: 16.13.2
V8: 10.0.139.17-electron.0
OS: Darwin arm64 21.5.0
```

**Additional context**

MacOS 12.4
Python 3.10.4

**Comment by erictraut** [comment]
I think pyright is correct here. 

It is the job of the type checker's constraint solver to determine whether there is a type that satisfies all of the constraints for a given call expression. Constraints can include an "upper bound" — i.e. the widest type allowed. The constraint solver widens or narrows the type to find a valid solution as long as it doesn't violate any of the constraints. An error is generated only if there is no possible solution that satisfies all of the constraints.
 
In your example above, there is a valid solution that meets all of the constraints for both `T` and `U`: `T = int` and `U = str | bool`.
 
The Python type system also includes the concept of a "constrained TypeVar" where you specify two or more types. The constraint solver then must pick one (but not multiple) of the constrained types.

Mypy appears to have inconsistent behavior here, and this looks to me like a bug in mypy.

Consider the following:

```python
from typing import Callable, TypeVar

R = TypeVar("R")

def cb1() -> bool:
    ...

def cb2() -> bool | str:
    ...

# Pyright and mypy agree that "() -> bool" is a
# subtype of "() -> bool | str"
x1: Callable[[], bool | str] = cb1

def func1(cb1: Callable[[], R], cb2: Callable[[], R]) -> R:
    ...

# Pyright and mypy are both able to solve R's type in this case.
v1 = func1(cb1, cb2)
reveal_type(v1)  # "bool | str"

def func2(v1: R, v2: R) -> R:
    ...

# Pyright and mypy are both able to solve R's type in this case,
# although mypy uses a join and pyright uses a union.
v2 = func2(1, "")
reveal_type(v2)  # pyright: "str | bool"; mypy: "object"

def func3(v1: R, cb1: Callable[[], R], cb2: Callable[[], R]) -> R:
    ...

# Pyright is able to solve R's type in this case, but mypy does not
# and generates an error.
v3 = func3("", cb1, cb2)
reveal_type(v3)  # pyright: "str | bool"; mypy: error
```

**Comment by andrewrosss** [comment]
This is super interesting. I really hadn't considered the _solving_ component of type checking. Looking at the examples through the lens of: "Is there a type that satisfies all the constraints?", inferring `v3` as `str | bool` makes perfect sense.

Thanks @erictraut for your thoughtful explanation and examples. Extremely helpful!
